src/device/cpu/riscv_rv32ima/tlb.c

```diff
@@ -0,0 +1,224 @@
+#include <string.h>
+
+#include "tlb.h"
+#include "../../../utils.h"
+
+typedef struct rv_tlb_entry {
+    sv32_pte_t pte;
+    uint32_t vpn;
+    unsigned asid;
+    bool valid;
+    bool global;
+} rv_tlb_entry_t;
+
+static inline size_t hash(size_t num){
+    // No hashing as of now
+    return num;
+}
+
+#define index_ktlb(tlb, index) \
+    ((tlb)->ktlb[hash((index)) % ((tlb)->ktlb_size)])
+
+#define index_mtlb(tlb, index) \
+    ((tlb)->mtlb[hash((index)) % ((tlb)->mtlb_size)])
+
+/** Caches a mapping into the TLB */
+extern void rv_tlb_add_mapping(rv_tlb_t* tlb, unsigned asid, uint32_t virt, sv32_pte_t pte, bool megapage, bool global){
+    
+    uint32_t vpn = virt >> RV_PAGESIZE;
```

Comment:
Probably the `vpn` variable could be moved into the `else` block?

Reply:
Yes it could.
Fixed in [d8210af](https://github.com/d-iii-s/msim/pull/36/commits/d8210afea32d8c76b88054151fed5de807a85116)